tools/test_xenial_upgrade.sh

```diff
@@ -0,0 +1,219 @@
+    set -e
+
+    h2 "New Status - verify esm-infra available but not enabled; esm-apps not visible"
+    lxc exec $name -- ua status
+
+    h2 "Attach - verify esm-infra automatically enabled; esm-apps not visible"
+    attach_ua lxc
+
+    h2 "Detaching before destruction"
+    lxc exec $name -- ua detach --assume-yes
+
+    teardown_container
+}
+
+
+function test_upgrade_with_livepatch_in_vm() {
+
+    setup_vm
+
+    h1 "Upgrade to UA 27 while old version has livepatch enabled"
+
+    h2 "Enable livepatch on old version"
+    set +x
+    echo "+ multipass exec $name -- ubuntu-advantage enable-livepatch \$LIVEPATCH_TOKEN"
+    multipass exec $name -- sudo ubuntu-advantage enable-livepatch $LIVEPATCH_TOKEN
+    set -x
+
+    h2 "Status before - old UA and livepatch say enabled"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ubuntu-advantage status
+
+    install_new_ua multipass
+
+    h2 "Status after upgrade - livepatch still enabled but new UA doesn't report it"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ua status
+
+    h2 "Attach - verify that livepatch is disabled and re-enabled"
+    attach_ua multipass
+
+    h2 "Status after attach - both livepatch and UA should say enabled"
+    multipass exec $name -- sudo canonical-livepatch status
+    multipass exec $name -- sudo ua status
+
+    h2 "Detaching before destruction"
+    multipass exec $name -- sudo ua detach --assume-yes
+
+    teardown_vm
+}
+
+function test_upgrade_with_fips_in_vm() {
+
+    setup_vm
+
+    h1 "Upgrade to UA 27 while old version has fips enabled"
+
+    h2 "Manual fips check says disabled (file doesn't exist)"
+    set +e
+    multipass exec $name -- sudo cat /proc/sys/crypto/fips_enabled
+    set -e
+
+    h2 "Enable fips on old version"
+    set +x
+    echo "+ multipass exec $name -- ubuntu-advantage enable-fips \$FIPS_CREDS"
+    multipass exec $name -- sudo ubuntu-advantage enable-fips $FIPS_CREDS
+    set -x
+
+    h2 "Reboot to finish fips activation"
+    set +e
+    multipass exec $name -- sudo reboot
+    set -e
+    sleep 20
+
+    h2 "Status before upgrade - old UA says fips is enabled, manual check agrees"
+    multipass exec $name -- sudo ubuntu-advantage status
+    multipass exec $name -- sudo cat /proc/sys/crypto/fips_enabled
+
+    install_new_ua multipass
+
+    h2 "Status after upgrade - new UA won't say anything is enabled, but a manual check still says fips is enabled"
+    multipass exec $name -- sudo ua status
+    multipass exec $name -- sudo cat /proc/sys/crypto/fips_enabled
+
+    h2 "Attach - only esm-infra will be auto-enabled"
+    attach_ua multipass
+
+    h2 "Status after attach - new UA will say fips is disabled and that livepatch is n/a"
+    multipass exec $name -- sudo ua status
+    multipass exec $name -- sudo cat /proc/sys/crypto/fips_enabled
+
```

Comment:
Can we add a check that our postinst has migrated oldclient FIPS APT sources to newclient and that apt update doesn't error
```bash
multipass exec $name -- sudo grep $FIPS_CREDS /etc/apt/sources.list.d/ubuntu-fips.list && echo "SUCCESS: found oldclient FIPS creds" || echo "FAIL: Migrated to new client creds"
multipass exec $name -- sudo apt update
```

Reply:
Yes good idea